dispatcher/dispatcher.go

```diff
@@ -0,0 +1,128 @@
+	ErrPersistence = esError("persistance failure")
+	// ErrTokenNotFound means the given token was not found in the dispatcher's reducer map.
+	ErrTokenNotFound = esError("token not found")
+)
+
+type esError string
+
+func (e esError) Error() string { return string(e) }
+
+// Dispatcher is used to dispatch events to registered reducers.
+//
+// This is different from generic pub-sub systems because reducers are not subscribed to particular events.
+// Every event is dispatched to every registered reducer. When a given reducer is registered, it returns a `token`,
+// which can be used to deregister the reducer later. The dispatcher wraps an underlying event store, which it uses to
+// persist events before dispatching to registered reducers. The whole system uses a mutex to prevent additional
+// dispatch calls from kicking off reducers until all in-flight reducers complete.
+type Dispatcher struct {
+	store    datastore.Datastore
+	reducers map[Token]dispatch.Reducer
+	lock     sync.Mutex
+}
+
+// NewDispatcher returns a new Dispatcher. This should only be called once in an application to ensure a singleton
+// dispatcher. While it is not enforced here, it is a good idea to a singleton pattern in your own code, for example:
+// 	var (
+// 		once sync.Once
+// 		instance *Dispatcher
+// 	)
+// 	var store = ...
+//
+//	func GetDispatcher() *Dispatcher {
+// 		once.Do(func() {
+// 			singleton = NewDispatcher(store)
+// 		})
+// 		return singleton
+// 	}
+func NewDispatcher(store datastore.Datastore) *Dispatcher {
+	return &Dispatcher{
+		store:    store,
+		reducers: make(map[Token]dispatch.Reducer),
+	}
+}
+
+// Register takes a reducer to be invoked with each dispatched event and returns a token for de-registration.
+func (d *Dispatcher) Register(reducer dispatch.Reducer) Token {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	lastID++
+	id := Token(lastID)
+	d.reducers[id] = reducer
+	return id
+}
+
+// Deregister removes a reducer based on its token. If the token is invalid (i.e. no associated reducer),
+// this is a no-op.
+func (d *Dispatcher) Deregister(token Token) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	if _, ok := d.reducers[token]; !ok {
+		return ErrTokenNotFound
+	}
+	delete(d.reducers, token)
+	return nil
+}
+
+// Dispatch dispatches a payload to all registered reducers. It returns a multierror object, which may contain
+// zero (nil) or more errors. Errors from reducer callbacks can be safely ignored and/or retried (and are prefixed
+// with "warning"), whereas errors due to event persistence are "critical" and will be prefixed as such.
+func (d *Dispatcher) Dispatch(event dispatch.Event) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	var result *multierror.Error
+	// Key format: <timestamp>/<entity-id>/<type>
+	// @todo: This is up for debate, its a 'fake' Event struct right now anyway
+	key := datastore.NewKey(string(event.Time())).ChildString(event.EntityID()).ChildString(event.Type())
```

Comment:
I would presume the `datastore` per application can be shared, so I think is a good idea to create a prefix key for the dispatcher. Like always starting with `/dispatcher` prefix in the keys to avoid a collision.

Reply:
This is a good idea. I hadn't thought that you'd share the store, but you totally could.

Reply:
Here's an example I did in the Keyboof if helps: https://github.com/textileio/go-textile-threads/blob/master/tstoreds/keybook.go